File: src/main/java/ru/igrey/dev/service/DealService.java
package ru.igrey.dev.service;

import lombok.RequiredArgsConstructor;
import ru.igrey.dev.dao.CrmTaskDao;
import ru.igrey.dev.dao.DealDao;
import ru.igrey.dev.dao.DocumentDao;
import ru.igrey.dev.dao.PXDao;
import ru.igrey.dev.domain.Deal;
import ru.igrey.dev.domain.Documents;
import ru.igrey.dev.entity.DealEntity;
import ru.igrey.dev.entity.RisasRequisitionEntity;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Created by sanasov on 10.04.2017.
 */
@RequiredArgsConstructor
public class DealService {
    private final DealDao dealDao;
    private final DocumentDao documentDao;
    private final CasService casService;
    private final ServiceHubService serviceHubService;
    private final CrmTaskDao crmTaskDao;
    private final PXDao pxDao;

    public Deal getDeal(Long dealId) {
        DealEntity dealEntity = dealDao.findById(dealId);
        if (dealEntity == null) return null;
        return new Deal(dealEntity.getId(),
                dealEntity.getDealStatusId(),
                dealEntity.getMikId(),
                dealEntity.getOfficeId(),
                dealEntity.getTypeId(),
                dealEntity.getArchived(),
                casService.casUser(dealEntity.getAuthorId()),
                crmTaskDao.crmTask(dealId),
                new Documents(
                        documentDao.findAllNotClientDocsByDealId(dealId),
                        documentDao.countUploadedDocuments(dealId)
                ),
                serviceHubService.getServices(dealId),
                pxDao.lastRequisition(dealId)
        );
    }

    public String getPhonesByDealIds(Set<Long> dealIds) {
        return "dealId    casId   phone\n" + dealDao.findByIds(dealIds).stream()
                .map(deal -> deal.getId() + "\t" + deal.getAuthorId() + "\t" + casService.casUser(deal.getAuthorId()).getPhone())
                .reduce((a, b) -> a + "\n" + b).orElse("");
    }

}

File: src/main/kotlin/ru/igrey/dev/service/DealService.kt
package ru.igrey.dev.service

import ru.igrey.dev.dao.CrmDao
import ru.igrey.dev.dao.DealDao
import ru.igrey.dev.dao.DocumentDao
import ru.igrey.dev.dao.PXDao
import ru.igrey.dev.domain.Deal
import ru.igrey.dev.domain.Documents
import ru.igrey.dev.entity.DealEntity

/**
 * Created by sanasov on 10.04.2017.
 */
class DealService(
    private val dealDao: DealDao,
    private val documentDao: DocumentDao,
    private val casService: CasService,
    private val serviceHubService: ServiceHubService,
    private val crmDao: CrmDao,
    private val pxDao: PXDao
) {

    fun getDeals(authorId: Long?): List<DealEntity> {
        return dealDao.findByAuthor(authorId)
    }

    fun getDeal(dealId: Long?): Deal? {
        val deal = dealDao.findById(dealId)
        val crmTasks = crmDao.crmTask(dealId)
        if (deal == null) return null
        return with(deal) {
            Deal(
                id,
                dealStatusId,
                casService.casUser(mikId),
                officeId,
                typeId,
                archived,
                casService.casUser(authorId),
                if (crmTasks.isEmpty()) "" else crmTasks[0],
                Documents(
                    documentDao.findAllNotClientDocsByDealId(dealId),
                    documentDao.countUploadedDocuments(dealId)!!
                ),
                serviceHubService.getServices(dealId!!),
                crmDao.schedule(dealId),
                pxDao.lastRequisition(dealId),
                dealDao.findLastHistoryById(dealId),
                createdTime
            )
        }
    }

    fun getPhonesByDealIds(dealIds: Set<Long>): String {
        return dealDao!!.findByIds(dealIds).stream()
            .map { (id, authorId) -> id.toString() + "\t" + casService!!.casUser(authorId)!!.toString() }
            .reduce { a, b -> a + "\n" + b }.orElse("")
    }
}
